lib/amqp_pool/pool.ex

```diff
@@ -0,0 +1,101 @@
+defmodule AMQPPool.Pool do
+  # See https://hexdocs.pm/elixir/Application.html
+  # for more information on OTP Applications
+  @moduledoc false
+
+  @default_pool_settings [
+    pool_size: 5,
+    max_overflow: 2
+  ]
+
+  def start_link(pool) do
+    Supervisor.start_link(__MODULE__, pool)
+  end
+
+  def init(pool) do
+    import Supervisor.Spec, warn: false
+
+    conn_worker_name = conn_worker_name(pool)
+
+    children = [
+      worker(AMQPPool.Connection, [amqp_connection_settings(pool), [name: conn_worker_name]]),
+      :poolboy.child_spec(pool, poolboy_config(pool), [conn_worker_name])
+    ]
+
+    # See https://hexdocs.pm/elixir/Supervisor.html
+    # for other strategies and supported options
+    opts = [strategy: :one_for_one, name: pool]
+    supervise(children, opts)
+  end
+
+  defp conn_worker_name(pool), do: Enum.join([AMQPPool.Connection, pool], ".") |> String.to_atom()
+
+  defp poolboy_config(pool) do
+    settings = Application.get_env(:amqp_pool, pool, @default_pool_settings)
+
+    [
+      {:name, {:local, pool}},
+      {:worker_module, AMQPPool.Channel},
+      {:size, settings[:pool_size]},
+      {:max_overflow, settings[:max_overflow]}
+    ]
+  end
+
+  defp amqp_connection_settings(pool) do
+    [
+      username: username(pool),
+      password: password(pool),
+      virtual_host: vhost(pool),
+      port: port(pool),
+      host: host(pool)
+    ]
+  end
+
+  def username(pool) do
```

Comment:
This pattern shows up 4 times, imo its worth to extract it into a function and reuse that.

Reply:
Correct! I could go for something like

```
def get_param(param_name, pool, transformation, default) do
  env_var = env_var(pool, param_name)
    param = param(pool, param_name)
    fetch_value(env_var, param, & &1, :undefined)
end
```

Reply:
Done!